feat(Nav): Convert Nav component to typescript, add Demos

jeff-phillips-18 opened the pull request:
**What**: 
-   Convert Nav component to typescript
-   Add Nav to Demo App
-   Add Nav integration tests

Fixes #2018

patternfly-build commented:
PatternFly-React preview: https://2005-pr-patternfly-react-patternfly.surge.sh

redallen commented on packages/patternfly-4/react-core/src/components/Nav/Nav.test.tsx:
```diff
@@ -20,8 +20,8 @@ beforeEach(() => {
 const context = {
-  onSelect: () => undefined,
-  onToggle: () => undefined
+  onSelect: ():void => undefined,
```
`() => undefined as any` is the pattern we're using so far.

redallen commented on packages/patternfly-4/react-core/src/components/Nav/Nav.tsx:
```diff
@@ -0,0 +1,94 @@
+  /** Additional classes added to the container */
+  className?: string;
+  /** Callback for updating when item selection changes */
+  onSelect?(selectedItem: {groupId: number | string; itemId: number | string; to:string, event: React.FormEvent<HTMLInputElement>}): void;
```
Sorry to nitpick, but please change to `onSelect?: (selectedItem: {groupId: number | string; itemId: number | string; to:string, event: React.FormEvent<HTMLInputElement>}) => void;` (use arrow functions for the types).

redallen commented on packages/patternfly-4/react-core/src/components/Nav/Nav.tsx:
```diff
@@ -0,0 +1,94 @@
+  }
+}
+
+export default Nav;
```
Remove default exports from all files.

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx:
```diff
@@ -0,0 +1,133 @@
+  /** Identifier to use for the section aria label */
+  id?: string;
+  /** allow consumer to optionally override this callback and manage expand state externally */
+  onExpand?(e: React.MouseEvent<HTMLLIElement, MouseEvent>, val: boolean): void;
```
Change to arrow function

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx:
```diff
@@ -0,0 +1,133 @@
+  }
+}
+
+export default NavExpandable;
```
Remove default export.

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavGroup.tsx:
```diff
@@ -1,29 +1,27 @@
+  };
 
-class NavGroup extends React.Component {
+  static componentType = 'NavGroup';
```
What's this needed for?

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx:
```diff
@@ -0,0 +1,91 @@
+  /** If true prevents the default anchor link action to occur. Set to true if you want to handle navigation yourself. */
+  preventDefault?: boolean;
+  /** Callback for item click */
+  onClick: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>, itemId: number | string, groupId: number | string) => void
```
You used arrow syntax here :tada:

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx:
```diff
@@ -0,0 +1,91 @@
+
+export interface NavItemProps extends Omit<React.HTMLProps<HTMLAnchorElement>,  'onClick'> {
+  /** Content rendered inside the nav item */
+  children?: React.ReactNode | string | null;
```
I think `React.ReactNode` covers the `| string | null`. Does removing it causes a problem?

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx:
```diff
@@ -0,0 +1,91 @@
+    groupId: null as string,
+    itemId: null as string,
+    preventDefault: false,
+    onClick: null as ((event: React.MouseEvent<HTMLAnchorElement, MouseEvent>, itemId: number | string, groupId: number | string) => void)
```
Create a type for this long thing and reference it in the props and here.

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx:
```diff
@@ -0,0 +1,91 @@
+    onClick: null as ((event: React.MouseEvent<HTMLAnchorElement, MouseEvent>, itemId: number | string, groupId: number | string) => void)
+  };
+
+  static componentType:string = 'NavItem';
```
What's this for?

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavItemSeparator.tsx:
```diff
@@ -0,0 +1,17 @@
+import styles from '@patternfly/patternfly/components/Nav/nav.css';
+import { css } from '@patternfly/react-styles';
+
+export class NavItemSeparator extends React.Component<React.HTMLProps<HTMLLIElement>> {
```
Can we make this back into a function component?

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavList.tsx:
```diff
@@ -0,0 +1,46 @@
+  /** Additional classes added to the list */
+  className?: string;
+  /** Indicates the list type. */
+  variant?: OneOf<typeof NavVariants, keyof typeof NavVariants>;
```
I think just `NavVariants` will do. If not, use `NavVariants | keyof typeof NavVariants`.

redallen commented on packages/patternfly-4/react-core/src/components/Nav/NavList.tsx:
```diff
@@ -0,0 +1,46 @@
+    className: ''
+  };
+
+  static componentType = 'NavList';
```
What's this do?

redallen commented on packages/patternfly-4/react-core/src/declarations.d.ts:
```diff
@@ -3,4 +3,18 @@
+
+// We use a JS babel plugin to resolve these imports that `tsc` doesn't
+// know about, so pretend that everything will work
+declare module '@patternfly/patternfly/layouts/*.css' {
```
I like this solution, but I think my PR is about to be merged (Alert) that just changes it to `declare module '@patternfly/patternfly/*/*.css'`, so you might want to go with that or deal with merge conflicts later.

redallen commented on packages/patternfly-4/react-integration/cypress/integration/nav.spec.ts:
```diff
@@ -0,0 +1,112 @@
+      expect(isCurrent).to.be.equal(index === 0);
+    });
+    cy.get('#default-link3').then((defaultLink3: JQuery<HTMLAnchorElement>) => {
+      cy.wrap(defaultLink3).click();
```
Wow, thanks for the awesome tests!

jeff-phillips-18 commented:
> Awesome work! Can we use `git mv` to rename the primary component files instead of deleting and recreating them? If we can do this, we'll retain the ability to trace the history of these modules going forward which is important for tracking down when certain changes were made, by who, and why. Looks like you've already done this for several components.

I did that. I tried my darnedest to keep then from being delete/add's but somehow they all got borked along the way (think it was a `git commit --amend`).

redallen commented:
@jeff-phillips-18 the way I've been doing it is to reset to the commit before your initial commit on this branch. So `git reset --soft ca245d37e856fd7dba9b364c26329af001cddd98~`. Then move your new .tsx files somewhere temporarily. Rename each .js -> .tsx. Commit. In the next commit change the contents of the tsx. It's a pain and requires two commits, but it keeps the history.

jeff-phillips-18 commented:
@redallen Have you noticed if it stays as a rename after a `Squash and merge`?

redallen commented:
No idea, I guess we'll find out how Github does that part when the first TS PR gets merged. That's actually a really good question.

seanforyou23 commented:
> Do you want me to try again?

No worries @jeff-phillips-18 - just where you can without getting blocked. I know we're converting a lot of components over the next while so more than anything I wanted to throw it out there so we can save as much history as possible and adopt this process where we can going forward. I should have gotten out in front of this before all these PR's popped up. That said, @redallen 's approach above should work. 

As far as squash and merge, I think if step zero of the working branch is to rename the file with git mv, then the subsequent commits operate on those files, logs should reflect the full history just fine. Thanks for looking into this guys, it's a big win for people inside and outside the project.

jeff-phillips-18 commented:
@redallen @seanforyou23 Updates made based on comments

jschuler commented on packages/patternfly-4/react-core/src/components/Nav/Nav.tsx:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  render() {
+    const { 'aria-label': ariaLabel, children, className, onSelect, ...props } = this.props;
```
i see you added onSelect here, maybe also spread out onToggle?

jschuler commented on packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx:
```diff
@@ -0,0 +1,129 @@
+  /** Boolean to programatically expand or collapse section */
+  isExpanded?: boolean;
+  /** Anything that can be rendered inside of the expandable list */
+  children?: React.ReactNode | string;
```
ReactNode should contain string, so this could just be ReactNode

jschuler commented:
I noticed that the page layout demos did not add preventDefault to the NavItem components. Adding it fixes the demo issues
https://github.com/jeff-phillips-18/patternfly-react/pull/1

jeff-phillips-18 commented:
Yes, the issue did exist prior to this PR. I added a change to auto-`preventDefault` for default nav item links if no `to` property is passed. Removed the unwanted `to` settings from the demo pages.

redallen commented on packages/patternfly-4/react-core/src/components/Nav/Nav.tsx:
```diff
@@ -0,0 +1,100 @@
+
+import { Omit } from '../../helpers/typeUtils';
+
+export type NavSelectClickHandler = (
```
Woohoo, this is readable!

jeff-phillips-18 commented:
@jschuler I believe all updates have been made
